src/server/worker/driver/driver.go

```diff
@@ -0,0 +1,1233 @@
+package driver
+
+import (
+	"bufio"
+	"bytes"
+	"context"
+	"encoding/json"
+	"fmt"
+	"io"
+	"io/ioutil"
+	"log"
+	"os"
+	"os/user"
+	"path/filepath"
+	"strconv"
+	"strings"
+	"sync"
+	"sync/atomic"
+	"syscall"
+	"time"
+	"unicode/utf8"
+
+	etcd "github.com/coreos/etcd/clientv3"
+	"github.com/gogo/protobuf/types"
+	"github.com/prometheus/client_golang/prometheus"
+	"gopkg.in/go-playground/webhooks.v5/github"
+	"gopkg.in/src-d/go-git.v4"
+	gitPlumbing "gopkg.in/src-d/go-git.v4/plumbing"
+
+	"github.com/pachyderm/pachyderm/src/client"
+	"github.com/pachyderm/pachyderm/src/client/enterprise"
+	"github.com/pachyderm/pachyderm/src/client/pfs"
+	"github.com/pachyderm/pachyderm/src/client/pkg/errors"
+	"github.com/pachyderm/pachyderm/src/client/pkg/grpcutil"
+	"github.com/pachyderm/pachyderm/src/client/pps"
+	col "github.com/pachyderm/pachyderm/src/server/pkg/collection"
+	"github.com/pachyderm/pachyderm/src/server/pkg/errutil"
+	"github.com/pachyderm/pachyderm/src/server/pkg/exec"
+	"github.com/pachyderm/pachyderm/src/server/pkg/hashtree"
+	"github.com/pachyderm/pachyderm/src/server/pkg/ppsconsts"
+	"github.com/pachyderm/pachyderm/src/server/pkg/ppsdb"
+	"github.com/pachyderm/pachyderm/src/server/pkg/ppsutil"
+	filesync "github.com/pachyderm/pachyderm/src/server/pkg/sync"
+	"github.com/pachyderm/pachyderm/src/server/pkg/uuid"
+	"github.com/pachyderm/pachyderm/src/server/pkg/work"
+	"github.com/pachyderm/pachyderm/src/server/worker/cache"
+	"github.com/pachyderm/pachyderm/src/server/worker/common"
+	"github.com/pachyderm/pachyderm/src/server/worker/logs"
+	"github.com/pachyderm/pachyderm/src/server/worker/stats"
+)
+
+const (
+	// The maximum number of concurrent download/upload operations
+	concurrency = 100
+)
+
+var (
+	errSpecialFile = errors.New("cannot upload special file")
+)
+
+func workNamespace(pipelineInfo *pps.PipelineInfo) string {
+	return fmt.Sprintf("/pipeline-%s/v%d", pipelineInfo.Pipeline.Name, pipelineInfo.Version)
+}
+
+// Driver provides an interface for common functions needed by worker code, and
+// captures the relevant objects necessary to provide these functions so that
+// users do not need to keep track of as many variables.  In addition, this
+// interface can be used to mock out external calls to make unit-testing
+// simpler.
+type Driver interface {
+	Jobs() col.Collection
+	Pipelines() col.Collection
+
+	NewTaskWorker() *work.Worker
+	NewTaskQueue() (*work.TaskQueue, error)
+
+	PipelineInfo() *pps.PipelineInfo
+
+	// The namespace that the worker is deployed in
+	Namespace() string
+
+	// Returns the path that will contain the input filesets for the job
+	InputDir() string
+
+	// Returns the pachd API client for the driver
+	PachClient() *client.APIClient
+
+	// Returns the number of workers to be used
+	GetExpectedNumWorkers() (uint64, error)
```

Comment:
I think you could probably just call this `ExpectedNumWorkers` without losing any clarity, also I'm generally a fan of using `int64` even if for numbers that can't be <0, eventually you wind up comparing them to signed integers and then you have to cast. If this is referencing values in protobufs then it might be more annoying to change this so feel free to skip. I regret my decision to use unsigned ints in protobufs.